Guard intrusive_ptr asserts with a flag.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31387 Guard intrusive_ptr asserts with a flag.**

These asserts add measurable performance overhead, but are really only
useful for debugging changes to intrusive_ptr itself, which is a
rare occurance. By putting them under a ifdef, we avoid overhead,
while still leaving them as a reference for when we are making changes
to intrusive_ptr.

Differential Revision: [D19148930](https://our.internmc.facebook.com/intern/diff/D19148930)

zdevito committed:
Guard intrusive_ptr asserts with a flag.

These asserts add measurable performance overhead, but are really only
useful for debugging changes to intrusive_ptr itself, which is a
rare occurance. By putting them under a ifdef, we avoid overhead,
while still leaving them as a reference for when we are making changes
to intrusive_ptr.

ngimel reviewed (approved):
LGTM

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 1bc9fc7:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=1bc9fc70736cf51e63b1366911375e8fdf34d0d4).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3991790) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76199877) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76199877))
<details>
<summary>
<code>Dec 17 23:48:53 RuntimeError: test_cpp_extensions failed!</code>
</summary>

```
Dec 17 23:48:53 /opt/conda/lib/python3.6/site-packages/torch/include/c10/util/intrusive_ptr.h:276:40:   required from 'c10::intrusive_ptr<T, NullType>& c10::intrusive_ptr<T, NullType>::operator=(const c10::intrusive_ptr<T, NullType>&) & [with TTarget = c10::VariableVersion::VersionCounter; NullType = c10::detail::intrusive_target_default_null_type<c10::VariableVersion::VersionCounter>]' 
Dec 17 23:48:53 /opt/conda/lib/python3.6/site-packages/torch/include/c10/core/TensorImpl.h:218:16:   required from here 
Dec 17 23:48:53 /opt/conda/lib/python3.6/site-packages/torch/include/c10/util/intrusive_ptr.h:193:14: error: unused variable 'new_refcount' [-Werror=unused-variable] 
Dec 17 23:48:53 cc1plus: all warnings being treated as errors 
Dec 17 23:48:53 error: command 'gcc' failed with exit status 1 
Dec 17 23:48:53 Traceback (most recent call last): 
Dec 17 23:48:53   File "test/run_test.py", line 456, in <module> 
Dec 17 23:48:53     main() 
Dec 17 23:48:53   File "test/run_test.py", line 449, in main 
Dec 17 23:48:53     raise RuntimeError(message) 
Dec 17 23:48:53 RuntimeError: test_cpp_extensions failed! 
Dec 17 23:48:53 =================== sccache compilation log =================== 
Dec 17 23:48:53 + cleanup 
Dec 17 23:48:53 + retcode=1 
Dec 17 23:48:53 + set +x 
c10::intrusive_ptr<T, NullType>&) & [with TTarget = c10::VariableVersion::VersionCounter; NullType = c10::detail::intrusive_target_default_null_type<c10::VariableVersion::VersionCounter>]\'\n/opt/conda/lib/python3.6/site-packages/torch/include/c10/core/TensorImpl.h:218:16:   required from here\n/opt/conda/lib/python3.6/site-packages/torch/include/c10/util/intrusive_ptr.h:193:14: error: unused variable \'new_refcount\' [-Werror=unused-variable]\ncc1plus: all warnings being treated as errors\n" } 
Dec 17 23:48:53  
Dec 17 23:48:53 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 17 23:48:53 Compile requests                  8 
Dec 17 23:48:53 Compile requests executed         7 
Dec 17 23:48:53 Cache hits                        0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zdevito committed:
Update on "Guard intrusive_ptr asserts with a flag."

Guard intrusive_ptr asserts with a flag.

These asserts add measurable performance overhead, but are really only
useful for debugging changes to intrusive_ptr itself, which is a
rare occurance. By putting them under a ifdef, we avoid overhead,
while still leaving them as a reference for when we are making changes
to intrusive_ptr.

gh-metadata: pytorch pytorch 31387 gh/zdevito/141/head

zdevito committed:
Update on "Guard intrusive_ptr asserts with a flag."

Guard intrusive_ptr asserts with a flag.

These asserts add measurable performance overhead, but are really only
useful for debugging changes to intrusive_ptr itself, which is a
rare occurance. By putting them under a ifdef, we avoid overhead,
while still leaving them as a reference for when we are making changes
to intrusive_ptr.

gh-metadata: pytorch pytorch 31387 gh/zdevito/141/head

zdevito committed:
Update on "Guard intrusive_ptr asserts with a flag."

Guard intrusive_ptr asserts with a flag.

These asserts add measurable performance overhead, but are really only
useful for debugging changes to intrusive_ptr itself, which is a
rare occurance. By putting them under a ifdef, we avoid overhead,
while still leaving them as a reference for when we are making changes
to intrusive_ptr.

gh-metadata: pytorch pytorch 31387 gh/zdevito/141/head
